### ISSUE TITLE
Slow completion for large Python packages when using the default pyls

### ISSUE THREAD

**Issue by jhettler** [krassowski/jupyterlab-lsp#272]
## Description

Autocompletion - especially of bigger packages like pandas is slow.

## Reproduce

1. Go to Jupyterlab
2. Write `import pandas as pd`
3. Write `pd.<tab>` 
4. It takes seconds in Jupyterlab, in classic notebook it is immediately autocompleted.

Is it usual behavior or there is something wrong configured?

## Expected behavior

Autocompletion finished nearly immediately :)

## Context

- Operating System and version:  _Linux b832f09e7bfa 5.4.0\-7626-generic #30\~1588169883\~20.04\~bbe668a\-Ubuntu SMP  x86_64 x86_64 x86_64 GNU/Linux_ - Docker
- Browser and version: _Chrome Version 81.0.4044.138 (Official Build) (64-bit)_
- JupyterLab version: _Jupyterlab version 2.1.0 (same behavior was with 2.1.2)_

<details><summary>Required: installed server extensions</summary>
<pre>
config dir: /home/jovyan/.jupyter
    jupyterlab_git  enabled 
config dir: /opt/conda/etc/jupyter
    jupyter_lsp  enabled 
    jupyterlab  enabled 
    jupyterlab_git  enabled 
    nbdime  enabled 
    - Validating...
      jupyterlab_git 0.20.0 OK
    - Validating...
      jupyter_lsp 0.8.0 OK
    - Validating...
      jupyterlab 2.1.0 OK
    - Validating...
      jupyterlab_git 0.20.0 OK
    - Validating...
      nbdime 2.0.0 OK
</pre>
</details>

<details><summary>Required: installed lab extensions</summary>
<pre>
   app dir: /opt/conda/share/jupyter/lab
        @aquirdturtle/collapsible_headings v2.2.0  enabled  OK
        @bokeh/jupyter_bokeh v2.0.0  enabled  OK
        @jupyter-widgets/jupyterlab-manager v2.0.0  enabled  OK
        @jupyterlab/git v0.20.0  enabled  OK
        @krassowski/jupyterlab-lsp v1.0.0  enabled  OK
        jupyter-matplotlib v0.7.2  enabled  OK
        jupyterlab-plotly v1.5.4  enabled  OK
        jupyterlab-python-file v0.4.0  enabled  OK
        nbdime-jupyterlab v2.0.0  enabled  OK
        plotlywidget v1.5.4  enabled  OK
        qgrid2 v1.1.3  enabled  OK
</pre>
</details>

<details><summary>Troubleshoot Output</summary>
<pre>
$PATH:
	/opt/conda/bin
	/usr/local/sbin
	/usr/local/bin
	/usr/sbin
	/usr/bin
	/sbin
	/bin

sys.path:
	/opt/conda/bin
	/opt/conda/lib/python37.zip
	/opt/conda/lib/python3.7
	/opt/conda/lib/python3.7/lib-dynload
	/opt/conda/lib/python3.7/site-packages

sys.executable:
	/opt/conda/bin/python

sys.version:
	3.7.6 | packaged by conda-forge | (default, Mar  5 2020, 15:27:18) 
	[GCC 7.3.0]

platform.platform():
	Linux-5.4.0-7626-generic-x86_64-with-debian-buster-sid

which -a jupyter:
	/opt/conda/bin/jupyter

pip list:
	Package                    Version            
	-------------------------- -------------------
	alembic                    1.4.1              
	asn1crypto                 1.3.0              
	astroid                    2.4.1              
	async-generator            1.10               
	attrs                      19.3.0             
	autopep8                   1.5.2              
	azure-common               1.1.25             
	azure-storage-blob         1.5.0              
	azure-storage-common       1.4.2              
	backcall                   0.1.0              
	beautifulsoup4             4.8.2              
	bleach                     3.1.3              
	blinker                    1.4                
	bokeh                      1.4.0              
	boto3                      1.13.12            
	botocore                   1.16.12            
	certifi                    2020.4.5.1         
	certipy                    0.1.3              
	cffi                       1.13.2             
	chardet                    3.0.4              
	click                      7.1.1              
	cloudpickle                1.3.0              
	colorama                   0.4.3              
	conda                      4.8.2              
	conda-package-handling     1.6.0              
	convertdate                2.2.1              
	cooked-input               0.5.2              
	coverage                   5.1                
	cryptography               2.8                
	cycler                     0.10.0             
	Cython                     0.29.15            
	cytoolz                    0.10.1             
	dask                       2.11.0             
	dateparser                 0.7.4              
	decorator                  4.4.2              
	defusedxml                 0.6.0              
	dill                       0.3.1.1            
	distributed                2.12.0             
	docutils                   0.15.2             
	entrypoints                0.3                
	enum34                     1.1.10             
	et-xmlfile                 1.0.1              
	flake8                     3.8.1              
	Flask                      1.1.2              
	Flask-Login                0.5.0              
	Flask-Mail                 0.9.1              
	Flask-Migrate              2.5.3              
	Flask-Principal            0.4.0              
	Flask-SQLAlchemy           2.4.1              
	fsspec                     0.6.2              
	future                     0.18.2             
	gitdb                      4.0.5              
	GitPython                  3.1.2              
	gunicorn                   20.0.4             
	h5py                       2.10.0             
	HeapDict                   1.0.1              
	holidays                   0.10.2             
	idna                       2.9                
	ijson                      2.6.1              
	imageio                    2.8.0              
	importlib-metadata         1.5.0              
	inflection                 0.4.0              
	ipykernel                  5.1.4              
	ipympl                     0.5.3              
	ipython                    7.13.0             
	ipython-genutils           0.2.0              
	ipython-sql                0.3.9              
	ipywidgets                 7.5.1              
	isort                      4.3.21             
	itsdangerous               1.1.0              
	jdcal                      1.4.1              
	jedi                       0.17.0             
	Jinja2                     2.11.1             
	jmespath                   0.10.0             
	joblib                     0.14.1             
	json5                      0.9.0              
	jsonschema                 3.2.0              
	jupyter-client             6.0.0              
	jupyter-core               4.6.3              
	jupyter-lsp                0.8.0              
	jupyter-telemetry          0.0.5              
	jupyterhub                 1.1.0              
	jupyterlab                 2.1.0              
	jupyterlab-git             0.20.0             
	jupyterlab-server          1.1.4              
	kiwisolver                 1.1.0              
	knowledge-repo             0.8.8              
	korean-lunar-calendar      0.2.1              
	lazy-object-proxy          1.4.3              
	llvmlite                   0.31.0             
	locket                     0.2.0              
	Mako                       1.1.0              
	Markdown                   3.2.2              
	MarkupSafe                 1.1.1              
	matplotlib                 3.2.1              
	mccabe                     0.6.1              
	mistune                    0.8.4              
	mpmath                     1.1.0              
	msgpack                    1.0.0              
	nbconvert                  5.6.1              
	nbdime                     2.0.0              
	nbformat                   5.0.4              
	networkx                   2.4                
	nose                       1.3.7              
	notebook                   6.0.3              
	numba                      0.48.0             
	numexpr                    2.7.1              
	numpy                      1.18.1             
	oauthlib                   3.0.1              
	olefile                    0.46               
	openpyxl                   3.0.3              
	oscrypto                   1.2.0              
	packaging                  20.1               
	pamela                     1.0.0              
	pandas                     1.0.2              
	pandocfilters              1.4.2              
	parso                      0.7.0              
	partd                      1.1.0              
	patsy                      0.5.1              
	pexpect                    4.8.0              
	pickleshare                0.7.5              
	Pillow                     7.0.0              
	pip                        20.0.2             
	plotly                     4.5.4              
	pluggy                     0.13.1             
	prettytable                0.7.2              
	prometheus-client          0.7.1              
	prompt-toolkit             3.0.4              
	protobuf                   3.11.4             
	psutil                     5.7.0              
	ptyprocess                 0.6.0              
	pycodestyle                2.6.0              
	pycosat                    0.6.3              
	pycparser                  2.20               
	pycryptodomex              3.9.7              
	pycurl                     7.43.0.5           
	pydocstyle                 5.0.2              
	pyflakes                   2.2.0              
	Pygments                   2.6.1              
	PyJWT                      1.7.1              
	pylint                     2.5.2              
	PyMeeus                    0.3.7              
	pyOpenSSL                  19.1.0             
	pyparsing                  2.4.6              
	PyPDF2                     1.26.0             
	pyrsistent                 0.15.7             
	PySocks                    1.7.1              
	python-dateutil            2.8.1              
	python-editor              1.0.4              
	python-json-logger         0.1.11             
	python-jsonrpc-server      0.3.4              
	python-language-server     0.32.0             
	pytz                       2019.3             
	PyWavelets                 1.1.1              
	PyYAML                     5.3                
	pyzmq                      19.0.0             
	qgrid                      1.3.1              
	regex                      2020.5.14          
	requests                   2.23.0             
	requests-oauthlib          1.3.0              
	retrying                   1.3.3              
	rope                       0.17.0             
	ruamel-yaml                0.15.80            
	ruamel.yaml                0.16.6             
	ruamel.yaml.clib           0.2.0              
	s3transfer                 0.3.3              
	scikit-image               0.16.2             
	scikit-learn               0.22.2.post1       
	scipy                      1.4.1              
	seaborn                    0.10.0             
	Send2Trash                 1.5.0              
	setuptools                 46.0.0.post20200311
	setuptools-git             1.2                
	six                        1.14.0             
	smmap                      3.0.4              
	snowballstemmer            2.0.0              
	snowflake-connector-python 2.2.6              
	snowflake-sqlalchemy       1.2.2              
	sortedcontainers           2.1.0              
	soupsieve                  1.9.4              
	SQLAlchemy                 1.3.15             
	sqlparse                   0.3.1              
	statsmodels                0.11.1             
	sympy                      1.5.1              
	tabulate                   0.8.7              
	tblib                      1.6.0              
	termcolor                  1.1.0              
	terminado                  0.8.3              
	testpath                   0.4.4              
	toml                       0.10.1             
	toolz                      0.10.0             
	tornado                    6.0.4              
	tqdm                       4.43.0             
	traitlets                  4.3.3              
	typed-ast                  1.4.1              
	tzlocal                    2.1                
	ujson                      1.35               
	urllib3                    1.25.7             
	validus                    0.3.0              
	veryprettytable            0.8.1              
	vincent                    0.4.4              
	Wand                       0.6.1              
	wcwidth                    0.1.8              
	webencodings               0.5.1              
	Werkzeug                   1.0.1              
	wheel                      0.34.2             
	widgetsnbextension         3.5.1              
	wrapt                      1.12.1             
	xlrd                       1.2.0              
	yapf                       0.30.0             
	zict                       2.0.0              
	zipp                       3.1.0

conda list:
	# packages in environment at /opt/conda:
	#
	# Name                    Version                   Build  Channel
	_libgcc_mutex             0.1                 conda_forge    conda-forge
	_openmp_mutex             4.5                       0_gnu    conda-forge
	alembic                   1.4.1                      py_0    conda-forge
	asn1crypto                1.3.0                    pypi_0    pypi
	astroid                   2.4.1                    pypi_0    pypi
	async_generator           1.10                       py_0    conda-forge
	attrs                     19.3.0                     py_0    conda-forge
	autopep8                  1.5.2                    pypi_0    pypi
	azure-common              1.1.25                   pypi_0    pypi
	azure-storage-blob        1.5.0                    pypi_0    pypi
	azure-storage-common      1.4.2                    pypi_0    pypi
	backcall                  0.1.0                      py_0    conda-forge
	beautifulsoup4            4.8.2                    py37_0    conda-forge
	blas                      2.14                   openblas    conda-forge
	bleach                    3.1.3              pyh8c360ce_0    conda-forge
	blinker                   1.4                        py_1    conda-forge
	bokeh                     1.4.0                    py37_0    conda-forge
	boto3                     1.13.12                  pypi_0    pypi
	botocore                  1.16.12                  pypi_0    pypi
	ca-certificates           2020.4.5.1           hecc5488_0    conda-forge
	certifi                   2020.4.5.1       py37hc8dfbb8_0    conda-forge
	certipy                   0.1.3                      py_0    conda-forge
	cffi                      1.13.2                   pypi_0    pypi
	chardet                   3.0.4           py37hc8dfbb8_1006    conda-forge
	click                     7.1.1              pyh8c360ce_0    conda-forge
	cloudpickle               1.3.0                      py_0    conda-forge
	colorama                  0.4.3                    pypi_0    pypi
	conda                     4.8.2                    py37_0    conda-forge
	conda-package-handling    1.6.0            py37h8f50634_2    conda-forge
	configurable-http-proxy   4.2.0           node13_he01fd0c_2    conda-forge
	convertdate               2.2.1                    pypi_0    pypi
	cooked-input              0.5.2                    pypi_0    pypi
	coverage                  5.1                      pypi_0    pypi
	cryptography              2.8              py37hb09aad4_2    conda-forge
	cycler                    0.10.0                     py_2    conda-forge
	cython                    0.29.15          py37h3340039_1    conda-forge
	cytoolz                   0.10.1           py37h516909a_0    conda-forge
	dask                      2.11.0                     py_0    conda-forge
	dask-core                 2.11.0                     py_0    conda-forge
	dateparser                0.7.4                    pypi_0    pypi
	decorator                 4.4.2                      py_0    conda-forge
	defusedxml                0.6.0                      py_0    conda-forge
	dill                      0.3.1.1          py37hc8dfbb8_1    conda-forge
	distributed               2.12.0                   py37_0    conda-forge
	docutils                  0.15.2                   pypi_0    pypi
	entrypoints               0.3             py37hc8dfbb8_1001    conda-forge
	enum34                    1.1.10                   pypi_0    pypi
	et-xmlfile                1.0.1                    pypi_0    pypi
	flake8                    3.8.1                    pypi_0    pypi
	flask                     1.1.2                    pypi_0    pypi
	flask-login               0.5.0                    pypi_0    pypi
	flask-mail                0.9.1                    pypi_0    pypi
	flask-migrate             2.5.3                    pypi_0    pypi
	flask-principal           0.4.0                    pypi_0    pypi
	flask-sqlalchemy          2.4.1                    pypi_0    pypi
	freetype                  2.10.0               he983fc9_1    conda-forge
	fsspec                    0.6.2                      py_0    conda-forge
	future                    0.18.2                   pypi_0    pypi
	giflib                    5.2.1                h516909a_2    conda-forge
	gitdb                     4.0.5                    pypi_0    pypi
	gitpython                 3.1.2                    pypi_0    pypi
	gunicorn                  20.0.4                   pypi_0    pypi
	h5py                      2.10.0          nompi_py37h513d04c_102    conda-forge
	hdf5                      1.10.5          nompi_h3c11f04_1104    conda-forge
	heapdict                  1.0.1                      py_0    conda-forge
	holidays                  0.10.2                   pypi_0    pypi
	icu                       64.2                 he1b5a44_1    conda-forge
	idna                      2.9                        py_1    conda-forge
	ijson                     2.6.1                    pypi_0    pypi
	imageio                   2.8.0                      py_0    conda-forge
	importlib-metadata        1.5.0            py37hc8dfbb8_1    conda-forge
	importlib_metadata        1.5.0                         1    conda-forge
	inflection                0.4.0                    pypi_0    pypi
	ipykernel                 5.1.4            py37h5ca1d4c_0    conda-forge
	ipympl                    0.5.3                      py_0    conda-forge
	ipython                   7.13.0           py37h5ca1d4c_0    conda-forge
	ipython-sql               0.3.9                    pypi_0    pypi
	ipython_genutils          0.2.0                      py_1    conda-forge
	ipywidgets                7.5.1                      py_0    conda-forge
	isort                     4.3.21                   pypi_0    pypi
	itsdangerous              1.1.0                    pypi_0    pypi
	jdcal                     1.4.1                    pypi_0    pypi
	jedi                      0.17.0                   pypi_0    pypi
	jinja2                    2.11.1                     py_0    conda-forge
	jmespath                  0.10.0                   pypi_0    pypi
	joblib                    0.14.1                     py_0    conda-forge
	jpeg                      9c                h14c3975_1001    conda-forge
	json5                     0.9.0                      py_0    conda-forge
	jsonschema                3.2.0            py37hc8dfbb8_1    conda-forge
	jupyter-lsp               0.8.0                    pypi_0    pypi
	jupyter_client            6.0.0                      py_0    conda-forge
	jupyter_core              4.6.3            py37hc8dfbb8_1    conda-forge
	jupyter_telemetry         0.0.5                      py_0    conda-forge
	jupyterhub                1.1.0                    py37_2    conda-forge
	jupyterhub-base           1.1.0                    py37_2    conda-forge
	jupyterlab                2.1.0                    pypi_0    pypi
	jupyterlab-git            0.20.0                   pypi_0    pypi
	jupyterlab-server         1.1.4                    pypi_0    pypi
	kiwisolver                1.1.0            py37h99015e2_1    conda-forge
	knowledge-repo            0.8.8                    pypi_0    pypi
	korean-lunar-calendar     0.2.1                    pypi_0    pypi
	krb5                      1.16.4               h2fd8d38_0    conda-forge
	lazy-object-proxy         1.4.3                    pypi_0    pypi
	ld_impl_linux-64          2.33.1               h53a641e_8    conda-forge
	libblas                   3.8.0               14_openblas    conda-forge
	libcblas                  3.8.0               14_openblas    conda-forge
	libcurl                   7.68.0               hda55be3_0    conda-forge
	libedit                   3.1.20170329      hf8c457e_1001    conda-forge
	libffi                    3.2.1             he1b5a44_1006    conda-forge
	libgcc-ng                 9.2.0                h24d8f2e_2    conda-forge
	libgfortran-ng            7.3.0                hdf63c60_5    conda-forge
	libgomp                   9.2.0                h24d8f2e_2    conda-forge
	liblapack                 3.8.0               14_openblas    conda-forge
	liblapacke                3.8.0               14_openblas    conda-forge
	libllvm8                  8.0.1                hc9558a2_0    conda-forge
	libopenblas               0.3.7                h5ec1e0e_6    conda-forge
	libpng                    1.6.37               hed695b0_0    conda-forge
	libprotobuf               3.11.4               h8b12597_0    conda-forge
	libsodium                 1.0.17               h516909a_0    conda-forge
	libssh2                   1.8.2                h22169c7_2    conda-forge
	libstdcxx-ng              9.2.0                hdf63c60_2    conda-forge
	libtiff                   4.1.0                hc7e4089_5    conda-forge
	libuv                     1.34.0               h516909a_0    conda-forge
	libwebp                   1.1.0                h56121f0_2    conda-forge
	libwebp-base              1.1.0                         2    conda-forge
	llvmlite                  0.31.0           py37h5202443_1    conda-forge
	locket                    0.2.0                      py_2    conda-forge
	lz4-c                     1.8.3             he1b5a44_1001    conda-forge
	mako                      1.1.0                      py_0    conda-forge
	markdown                  3.2.2                    pypi_0    pypi
	markupsafe                1.1.1            py37h8f50634_1    conda-forge
	matplotlib                3.2.1                    pypi_0    pypi
	mccabe                    0.6.1                    pypi_0    pypi
	mistune                   0.8.4           py37h516909a_1000    conda-forge
	mpmath                    1.1.0                      py_0    conda-forge
	msgpack-python            1.0.0            py37h99015e2_1    conda-forge
	nbconvert                 5.6.1                    py37_0    conda-forge
	nbdime                    2.0.0                    pypi_0    pypi
	nbformat                  5.0.4                      py_0    conda-forge
	ncurses                   6.1               hf484d3e_1002    conda-forge
	networkx                  2.4                        py_1    conda-forge
	nodejs                    13.10.1              hf5d1a2b_0    conda-forge
	nose                      1.3.7                    pypi_0    pypi
	notebook                  6.0.3                    py37_0    conda-forge
	numba                     0.48.0           py37hb3f55d8_0    conda-forge
	numexpr                   2.7.1            py37hb3f55d8_0    conda-forge
	numpy                     1.18.1           py37h95a1406_0    conda-forge
	oauthlib                  3.0.1                      py_0    conda-forge
	olefile                   0.46                       py_0    conda-forge
	openpyxl                  3.0.3                    pypi_0    pypi
	openssl                   1.1.1g               h516909a_0    conda-forge
	oscrypto                  1.2.0                    pypi_0    pypi
	packaging                 20.1                       py_0    conda-forge
	pamela                    1.0.0                      py_0    conda-forge
	pandas                    1.0.2            py37h0da4684_0    conda-forge
	pandoc                    2.9.2                         0    conda-forge
	pandocfilters             1.4.2                      py_1    conda-forge
	parso                     0.7.0                    pypi_0    pypi
	partd                     1.1.0                      py_0    conda-forge
	patsy                     0.5.1                      py_0    conda-forge
	pexpect                   4.8.0            py37hc8dfbb8_1    conda-forge
	pickleshare               0.7.5           py37hc8dfbb8_1001    conda-forge
	pillow                    7.0.0            py37h718be6c_1    conda-forge
	pip                       20.0.2                     py_2    conda-forge
	plotly                    4.5.4                    pypi_0    pypi
	pluggy                    0.13.1                   pypi_0    pypi
	prettytable               0.7.2                    pypi_0    pypi
	prometheus_client         0.7.1                      py_0    conda-forge
	prompt-toolkit            3.0.4                      py_0    conda-forge
	prompt_toolkit            3.0.4                         0    conda-forge
	protobuf                  3.11.4           py37he1b5a44_0    conda-forge
	psutil                    5.7.0            py37h8f50634_1    conda-forge
	ptyprocess                0.6.0                   py_1001    conda-forge
	pycodestyle               2.6.0                    pypi_0    pypi
	pycosat                   0.6.3           py37h8f50634_1004    conda-forge
	pycparser                 2.20                       py_0    conda-forge
	pycryptodomex             3.9.7                    pypi_0    pypi
	pycurl                    7.43.0.5         py37h16ce93b_0    conda-forge
	pydocstyle                5.0.2                    pypi_0    pypi
	pyflakes                  2.2.0                    pypi_0    pypi
	pygments                  2.6.1                      py_0    conda-forge
	pyjwt                     1.7.1                      py_0    conda-forge
	pylint                    2.5.2                    pypi_0    pypi
	pymeeus                   0.3.7                    pypi_0    pypi
	pyopenssl                 19.1.0                     py_1    conda-forge
	pyparsing                 2.4.6                      py_0    conda-forge
	pypdf2                    1.26.0                   pypi_0    pypi
	pyrsistent                0.15.7           py37h8f50634_1    conda-forge
	pysocks                   1.7.1            py37hc8dfbb8_1    conda-forge
	python                    3.7.6           h357f687_4_cpython    conda-forge
	python-dateutil           2.8.1                      py_0    conda-forge
	python-editor             1.0.4                      py_0    conda-forge
	python-json-logger        0.1.11                     py_0    conda-forge
	python-jsonrpc-server     0.3.4                    pypi_0    pypi
	python-language-server    0.32.0                   pypi_0    pypi
	python_abi                3.7                     1_cp37m    conda-forge
	pytz                      2019.3                     py_0    conda-forge
	pywavelets                1.1.1            py37hc1659b7_0    conda-forge
	pyyaml                    5.3              py37h8f50634_1    conda-forge
	pyzmq                     19.0.0           py37hac76be4_1    conda-forge
	qgrid                     1.3.1                    pypi_0    pypi
	readline                  8.0                  hf8c457e_0    conda-forge
	regex                     2020.5.14                pypi_0    pypi
	requests                  2.23.0             pyh8c360ce_2    conda-forge
	requests-oauthlib         1.3.0                    pypi_0    pypi
	retrying                  1.3.3                    pypi_0    pypi
	rope                      0.17.0                   pypi_0    pypi
	ruamel.yaml               0.16.6           py37h8f50634_1    conda-forge
	ruamel.yaml.clib          0.2.0            py37h8f50634_1    conda-forge
	ruamel_yaml               0.15.80         py37h8f50634_1001    conda-forge
	s3transfer                0.3.3                    pypi_0    pypi
	scikit-image              0.16.2           py37hb3f55d8_0    conda-forge
	scikit-learn              0.22.2.post1     py37hcdab131_0    conda-forge
	scipy                     1.4.1            py37h921218d_0    conda-forge
	seaborn                   0.10.0                     py_1    conda-forge
	send2trash                1.5.0                      py_0    conda-forge
	setuptools                46.0.0           py37hc8dfbb8_2    conda-forge
	setuptools-git            1.2                      pypi_0    pypi
	six                       1.14.0                     py_1    conda-forge
	smmap                     3.0.4                    pypi_0    pypi
	snowballstemmer           2.0.0                    pypi_0    pypi
	snowflake-connector-python 2.2.6                    pypi_0    pypi
	snowflake-sqlalchemy      1.2.2                    pypi_0    pypi
	sortedcontainers          2.1.0                      py_0    conda-forge
	soupsieve                 1.9.4            py37hc8dfbb8_1    conda-forge
	sqlalchemy                1.3.15           py37h8f50634_1    conda-forge
	sqlite                    3.30.1               hcee41ef_0    conda-forge
	sqlparse                  0.3.1                    pypi_0    pypi
	statsmodels               0.11.1           py37h8f50634_1    conda-forge
	sympy                     1.5.1            py37hc8dfbb8_2    conda-forge
	tabulate                  0.8.7                    pypi_0    pypi
	tblib                     1.6.0                      py_0    conda-forge
	termcolor                 1.1.0                    pypi_0    pypi
	terminado                 0.8.3            py37hc8dfbb8_1    conda-forge
	testpath                  0.4.4                      py_0    conda-forge
	tini                      0.18.0            h14c3975_1001    conda-forge
	tk                        8.6.10               hed695b0_0    conda-forge
	toml                      0.10.1                   pypi_0    pypi
	toolz                     0.10.0                     py_0    conda-forge
	tornado                   6.0.4            py37h8f50634_1    conda-forge
	tqdm                      4.43.0                     py_0    conda-forge
	traitlets                 4.3.3            py37hc8dfbb8_1    conda-forge
	typed-ast                 1.4.1                    pypi_0    pypi
	tzlocal                   2.1                      pypi_0    pypi
	ujson                     1.35                     pypi_0    pypi
	urllib3                   1.25.7           py37hc8dfbb8_1    conda-forge
	validus                   0.3.0                    pypi_0    pypi
	veryprettytable           0.8.1                    pypi_0    pypi
	vincent                   0.4.4                      py_1    conda-forge
	wand                      0.6.1                    pypi_0    pypi
	wcwidth                   0.1.8                      py_0    conda-forge
	webencodings              0.5.1                      py_1    conda-forge
	werkzeug                  1.0.1                    pypi_0    pypi
	wheel                     0.34.2                     py_1    conda-forge
	widgetsnbextension        3.5.1                    py37_0    conda-forge
	wrapt                     1.12.1                   pypi_0    pypi
	xlrd                      1.2.0                      py_0    conda-forge
	xz                        5.2.4             h14c3975_1001    conda-forge
	yaml                      0.2.2                h516909a_1    conda-forge
	yapf                      0.30.0                   pypi_0    pypi
	zeromq                    4.3.2                he1b5a44_2    conda-forge
	zict                      2.0.0                      py_0    conda-forge
	zipp                      3.1.0                      py_0    conda-forge
	zlib                      1.2.11            h516909a_1006    conda-forge
	zstd                      1.4.4                h3b9ef0a_1    conda-forge
</pre>
</details>

<details><summary>Command Line Output</summary>
<pre>
[D 19:50:19.900 LabApp] Searching ['/home/jovyan/.jupyter', '/home/jovyan/.jupyter', '/opt/conda/etc/jupyter', '/usr/local/etc/jupyter', '/etc/jupyter'] for config files
[D 19:50:19.900 LabApp] Looking for jupyter_config in /etc/jupyter
[D 19:50:19.900 LabApp] Looking for jupyter_config in /usr/local/etc/jupyter
[D 19:50:19.901 LabApp] Looking for jupyter_config in /opt/conda/etc/jupyter
[D 19:50:19.901 LabApp] Looking for jupyter_config in /home/jovyan/.jupyter
[D 19:50:19.901 LabApp] Looking for jupyter_config in /home/jovyan/.jupyter
[D 19:50:19.904 LabApp] Looking for jupyter_notebook_config in /etc/jupyter
[D 19:50:19.904 LabApp] Loaded config file: /etc/jupyter/jupyter_notebook_config.py
[D 19:50:19.905 LabApp] Looking for jupyter_notebook_config in /usr/local/etc/jupyter
[D 19:50:19.905 LabApp] Looking for jupyter_notebook_config in /opt/conda/etc/jupyter
[D 19:50:19.905 LabApp] Looking for jupyter_notebook_config in /home/jovyan/.jupyter
[D 19:50:19.906 LabApp] Loaded config file: /home/jovyan/.jupyter/jupyter_notebook_config.py
[D 19:50:19.906 LabApp] Loaded config file: /home/jovyan/.jupyter/jupyter_notebook_config.json
[D 19:50:19.906 LabApp] Looking for jupyter_notebook_config in /home/jovyan/.jupyter
[D 19:50:19.906 LabApp] Loaded config file: /home/jovyan/.jupyter/jupyter_notebook_config.py
[D 19:50:19.907 LabApp] Loaded config file: /home/jovyan/.jupyter/jupyter_notebook_config.json
[D 19:50:19.919 LabApp] Paths used for configuration of jupyter_notebook_config: 
        /etc/jupyter/jupyter_notebook_config.json
[D 19:50:19.920 LabApp] Paths used for configuration of jupyter_notebook_config: 
        /usr/local/etc/jupyter/jupyter_notebook_config.json
[D 19:50:19.921 LabApp] Paths used for configuration of jupyter_notebook_config: 
        /opt/conda/etc/jupyter/jupyter_notebook_config.d/jupyter-lsp-serverextension.json
        /opt/conda/etc/jupyter/jupyter_notebook_config.d/jupyterlab.json
        /opt/conda/etc/jupyter/jupyter_notebook_config.d/jupyterlab_git.json
        /opt/conda/etc/jupyter/jupyter_notebook_config.d/nbdime.json
        /opt/conda/etc/jupyter/jupyter_notebook_config.json
[D 19:50:19.922 LabApp] Paths used for configuration of jupyter_notebook_config: 
        /home/jovyan/.jupyter/jupyter_notebook_config.json
[I 19:50:20.255 LabApp] The port 8888 is already in use, trying another port.
[D 19:50:20.312 LabApp] Checking for /home/jovyan/.jupyter/node_modules/bash-language-server/bin/main.js
[D 19:50:20.312 LabApp] Checking for /opt/conda/share/jupyter/lab/staging/node_modules/bash-language-server/bin/main.js
[D 19:50:20.313 LabApp] Checking for /opt/conda/lib/node_modules/bash-language-server/bin/main.js
[D 19:50:20.313 LabApp] Checking for /opt/conda/node_modules/bash-language-server/bin/main.js
[D 19:50:20.313 LabApp] bash-language-server/bin/main.js not found in node_modules of [PosixPath('/home/jovyan/.jupyter'), PosixPath('/opt/conda/share/jupyter/lab/staging'), PosixPath('/opt/conda/lib'), PosixPath('/opt/conda')]
[D 19:50:20.313 LabApp] Checking for /home/jovyan/.jupyter/node_modules/dockerfile-language-server-nodejs/lib/server.js
[D 19:50:20.313 LabApp] Checking for /opt/conda/share/jupyter/lab/staging/node_modules/dockerfile-language-server-nodejs/lib/server.js
[D 19:50:20.313 LabApp] Checking for /opt/conda/lib/node_modules/dockerfile-language-server-nodejs/lib/server.js
[D 19:50:20.314 LabApp] Checking for /opt/conda/node_modules/dockerfile-language-server-nodejs/lib/server.js
[D 19:50:20.314 LabApp] dockerfile-language-server-nodejs/lib/server.js not found in node_modules of [PosixPath('/home/jovyan/.jupyter'), PosixPath('/opt/conda/share/jupyter/lab/staging'), PosixPath('/opt/conda/lib'), PosixPath('/opt/conda')]
[D 19:50:20.314 LabApp] Checking for /home/jovyan/.jupyter/node_modules/javascript-typescript-langserver/lib/language-server-stdio.js
[D 19:50:20.314 LabApp] Checking for /opt/conda/share/jupyter/lab/staging/node_modules/javascript-typescript-langserver/lib/language-server-stdio.js
[D 19:50:20.314 LabApp] Checking for /opt/conda/lib/node_modules/javascript-typescript-langserver/lib/language-server-stdio.js
[D 19:50:20.315 LabApp] Checking for /opt/conda/node_modules/javascript-typescript-langserver/lib/language-server-stdio.js
[D 19:50:20.315 LabApp] javascript-typescript-langserver/lib/language-server-stdio.js not found in node_modules of [PosixPath('/home/jovyan/.jupyter'), PosixPath('/opt/conda/share/jupyter/lab/staging'), PosixPath('/opt/conda/lib'), PosixPath('/opt/conda')]
[D 19:50:20.318 LabApp] Checking for /home/jovyan/.jupyter/node_modules/unified-language-server/src/server.js
[D 19:50:20.318 LabApp] Checking for /opt/conda/share/jupyter/lab/staging/node_modules/unified-language-server/src/server.js
[D 19:50:20.319 LabApp] Checking for /opt/conda/lib/node_modules/unified-language-server/src/server.js
[D 19:50:20.319 LabApp] Checking for /opt/conda/node_modules/unified-language-server/src/server.js
[D 19:50:20.319 LabApp] unified-language-server/src/server.js not found in node_modules of [PosixPath('/home/jovyan/.jupyter'), PosixPath('/opt/conda/share/jupyter/lab/staging'), PosixPath('/opt/conda/lib'), PosixPath('/opt/conda')]
[D 19:50:20.319 LabApp] Checking for /home/jovyan/.jupyter/node_modules/vscode-css-languageserver-bin/cssServerMain.js
[D 19:50:20.319 LabApp] Checking for /opt/conda/share/jupyter/lab/staging/node_modules/vscode-css-languageserver-bin/cssServerMain.js
[D 19:50:20.320 LabApp] Checking for /opt/conda/lib/node_modules/vscode-css-languageserver-bin/cssServerMain.js
[D 19:50:20.320 LabApp] Checking for /opt/conda/node_modules/vscode-css-languageserver-bin/cssServerMain.js
[D 19:50:20.320 LabApp] vscode-css-languageserver-bin/cssServerMain.js not found in node_modules of [PosixPath('/home/jovyan/.jupyter'), PosixPath('/opt/conda/share/jupyter/lab/staging'), PosixPath('/opt/conda/lib'), PosixPath('/opt/conda')]
[D 19:50:20.320 LabApp] Checking for /home/jovyan/.jupyter/node_modules/vscode-html-languageserver-bin/htmlServerMain.js
[D 19:50:20.320 LabApp] Checking for /opt/conda/share/jupyter/lab/staging/node_modules/vscode-html-languageserver-bin/htmlServerMain.js
[D 19:50:20.320 LabApp] Checking for /opt/conda/lib/node_modules/vscode-html-languageserver-bin/htmlServerMain.js
[D 19:50:20.321 LabApp] Checking for /opt/conda/node_modules/vscode-html-languageserver-bin/htmlServerMain.js
[D 19:50:20.321 LabApp] vscode-html-languageserver-bin/htmlServerMain.js not found in node_modules of [PosixPath('/home/jovyan/.jupyter'), PosixPath('/opt/conda/share/jupyter/lab/staging'), PosixPath('/opt/conda/lib'), PosixPath('/opt/conda')]
[D 19:50:20.321 LabApp] Checking for /home/jovyan/.jupyter/node_modules/vscode-json-languageserver-bin/jsonServerMain.js
[D 19:50:20.321 LabApp] Checking for /opt/conda/share/jupyter/lab/staging/node_modules/vscode-json-languageserver-bin/jsonServerMain.js
[D 19:50:20.321 LabApp] Checking for /opt/conda/lib/node_modules/vscode-json-languageserver-bin/jsonServerMain.js
[D 19:50:20.321 LabApp] Checking for /opt/conda/node_modules/vscode-json-languageserver-bin/jsonServerMain.js
[D 19:50:20.322 LabApp] vscode-json-languageserver-bin/jsonServerMain.js not found in node_modules of [PosixPath('/home/jovyan/.jupyter'), PosixPath('/opt/conda/share/jupyter/lab/staging'), PosixPath('/opt/conda/lib'), PosixPath('/opt/conda')]
[D 19:50:20.322 LabApp] Checking for /home/jovyan/.jupyter/node_modules/yaml-language-server/bin/yaml-language-server
[D 19:50:20.322 LabApp] Checking for /opt/conda/share/jupyter/lab/staging/node_modules/yaml-language-server/bin/yaml-language-server
[D 19:50:20.322 LabApp] Checking for /opt/conda/lib/node_modules/yaml-language-server/bin/yaml-language-server
[D 19:50:20.322 LabApp] Checking for /opt/conda/node_modules/yaml-language-server/bin/yaml-language-server
[D 19:50:20.323 LabApp] yaml-language-server/bin/yaml-language-server not found in node_modules of [PosixPath('/home/jovyan/.jupyter'), PosixPath('/opt/conda/share/jupyter/lab/staging'), PosixPath('/opt/conda/lib'), PosixPath('/opt/conda')]
[D 19:50:20.440 LabApp] [lsp] rootUri will be file:///home/jovyan/.jupyter
[D 19:50:20.440 LabApp] [lsp] virtualDocumentsUri will be file:///home/jovyan/.jupyter/.virtual_documents
[D 19:50:20.442 LabApp] [lsp] The following Language Servers will be available: {
      "pyls": {
        "argv": [
          "/opt/conda/bin/pyls"
        ],
        "config_schema": {
          "properties": {
            "pyls.configurationSources": {
              "default": [
                "pycodestyle"
              ],
              "description": "List of configuration sources to use.",
              "items": {
                "enum": [
                  "pycodestyle",
                  "pyflakes"
                ],
                "type": "string"
              },
              "type": "array",
              "uniqueItems": true
            },
            "pyls.executable": {
              "default": "pyls",
              "description": "Language server executable",
              "type": "string"
            },
            "pyls.plugins.jedi_completion.enabled": {
              "default": true,
              "description": "Enable or disable the plugin.",
              "type": "boolean"
            },
            "pyls.plugins.jedi_completion.include_params": {
              "default": true,
              "description": "Auto-completes methods and classes with tabstops for each parameter.",
              "type": "boolean"
            },
            "pyls.plugins.jedi_definition.enabled": {
              "default": true,
              "description": "Enable or disable the plugin.",
              "type": "boolean"
            },
            "pyls.plugins.jedi_definition.follow_builtin_imports": {
              "default": true,
              "description": "If follow_imports is True will decide if it follow builtin imports.",
              "type": "boolean"
            },
            "pyls.plugins.jedi_definition.follow_imports": {
              "default": true,
              "description": "The goto call will follow imports.",
              "type": "boolean"
            },
            "pyls.plugins.jedi_hover.enabled": {
              "default": true,
              "description": "Enable or disable the plugin.",
              "type": "boolean"
            },
            "pyls.plugins.jedi_references.enabled": {
              "default": true,
              "description": "Enable or disable the plugin.",
              "type": "boolean"
            },
            "pyls.plugins.jedi_signature_help.enabled": {
              "default": true,
              "description": "Enable or disable the plugin.",
              "type": "boolean"
            },
            "pyls.plugins.jedi_symbols.all_scopes": {
              "default": true,
              "description": "If True lists the names of all scopes instead of only the module namespace.",
              "type": "boolean"
            },
            "pyls.plugins.jedi_symbols.enabled": {
              "default": true,
              "description": "Enable or disable the plugin.",
              "type": "boolean"
            },
            "pyls.plugins.mccabe.enabled": {
              "default": true,
              "description": "Enable or disable the plugin.",
              "type": "boolean"
            },
            "pyls.plugins.mccabe.threshold": {
              "default": 15,
              "description": "The minimum threshold that triggers warnings about cyclomatic complexity.",
              "type": "number"
            },
            "pyls.plugins.preload.enabled": {
              "default": true,
              "description": "Enable or disable the plugin.",
              "type": "boolean"
            },
            "pyls.plugins.preload.modules": {
              "default": null,
              "description": "List of modules to import on startup",
              "items": {
                "type": "string"
              },
              "type": "array",
              "uniqueItems": true
            },
            "pyls.plugins.pycodestyle.enabled": {
              "default": true,
              "description": "Enable or disable the plugin.",
              "type": "boolean"
            },
            "pyls.plugins.pycodestyle.exclude": {
              "default": null,
              "description": "Exclude files or directories which match these patterns.",
              "items": {
                "type": "string"
              },
              "type": "array",
              "uniqueItems": true
            },
            "pyls.plugins.pycodestyle.filename": {
              "default": null,
              "description": "When parsing directories, only check filenames matching these patterns.",
              "items": {
                "type": "string"
              },
              "type": "array",
              "uniqueItems": true
            },
            "pyls.plugins.pycodestyle.hangClosing": {
              "default": null,
              "description": "Hang closing bracket instead of matching indentation of opening bracket's line.",
              "type": "boolean"
            },
            "pyls.plugins.pycodestyle.ignore": {
              "default": null,
              "description": "Ignore errors and warnings",
              "items": {
                "type": "string"
              },
              "type": "array",
              "uniqueItems": true
            },
            "pyls.plugins.pycodestyle.maxLineLength": {
              "default": null,
              "description": "Set maximum allowed line length.",
              "type": "number"
            },
            "pyls.plugins.pycodestyle.select": {
              "default": null,
              "description": "Select errors and warnings",
              "items": {
                "type": "string"
              },
              "type": "array",
              "uniqueItems": true
            },
            "pyls.plugins.pydocstyle.addIgnore": {
              "default": null,
              "description": "Ignore errors and warnings in addition to the specified convention.",
              "items": {
                "type": "string"
              },
              "type": "array",
              "uniqueItems": true
            },
            "pyls.plugins.pydocstyle.addSelect": {
              "default": null,
              "description": "Select errors and warnings in addition to the specified convention.",
              "items": {
                "type": "string"
              },
              "type": "array",
              "uniqueItems": true
            },
            "pyls.plugins.pydocstyle.convention": {
              "default": null,
              "description": "Choose the basic list of checked errors by specifying an existing convention.",
              "enum": [
                "pep257",
                "numpy"
              ],
              "type": "string"
            },
            "pyls.plugins.pydocstyle.enabled": {
              "default": false,
              "description": "Enable or disable the plugin.",
              "type": "boolean"
            },
            "pyls.plugins.pydocstyle.ignore": {
              "default": null,
              "description": "Ignore errors and warnings",
              "items": {
                "type": "string"
              },
              "type": "array",
              "uniqueItems": true
            },
            "pyls.plugins.pydocstyle.match": {
              "default": "(?!test_).*\\.py",
              "description": "Check only files that exactly match the given regular expression; default is to match files that don't start with 'test_' but end with '.py'.",
              "type": "string"
            },
            "pyls.plugins.pydocstyle.matchDir": {
              "default": "[^\\.].*",
              "description": "Search only dirs that exactly match the given regular expression; default is to match dirs which do not begin with a dot.",
              "type": "string"
            },
            "pyls.plugins.pydocstyle.select": {
              "default": null,
              "description": "Select errors and warnings",
              "items": {
                "type": "string"
              },
              "type": "array",
              "uniqueItems": true
            },
            "pyls.plugins.pyflakes.enabled": {
              "default": true,
              "description": "Enable or disable the plugin.",
              "type": "boolean"
            },
            "pyls.plugins.pylint.args": {
              "default": null,
              "description": "Arguments to pass to pylint.",
              "items": {
                "type": "string"
              },
              "type": "array",
              "uniqueItems": false
            },
            "pyls.plugins.pylint.enabled": {
              "default": true,
              "description": "Enable or disable the plugin.",
              "type": "boolean"
            },
            "pyls.plugins.rope_completion.enabled": {
              "default": true,
              "description": "Enable or disable the plugin.",
              "type": "boolean"
            },
            "pyls.plugins.yapf.enabled": {
              "default": true,
              "description": "Enable or disable the plugin.",
              "type": "boolean"
            },
            "pyls.rope.extensionModules": {
              "default": null,
              "description": "Builtin and c-extension modules that are allowed to be imported and inspected by rope.",
              "type": "string"
            },
            "pyls.rope.ropeFolder": {
              "default": null,
              "description": "The name of the folder in which rope stores project configurations and data.  Pass `null` for not using such a folder at all.",
              "items": {
                "type": "string"
              },
              "type": "array",
              "uniqueItems": true
            }
          },
          "title": "Python Language Server Configuration",
          "type": "object"
        },
        "display_name": "pyls",
        "env": {
          "PYTHONUNBUFFERED": "1"
        },
        "extend": [
          {
            "display_name": "pyls-mypy",
            "install": {
              "pip": "pip install pyls-mypy"
            }
          },
          {
            "display_name": "pyls-black",
            "install": {
              "pip": "pip install pyls-black"
            }
          },
          {
            "display_name": "pyls-isort",
            "install": {
              "pip": "pip install pyls-isort"
            }
          }
        ],
        "install": {
          "conda": "conda install -c conda-forge python-language-server",
          "pip": "pip install python-language-server[all]"
        },
        "languages": [
          "python"
        ],
        "mime_types": [
          "text/python",
          "text/x-ipython"
        ],
        "urls": {
          "home": "https://github.com/palantir/python-language-server",
          "issues": "https://github.com/palantir/python-language-server/issues"
        },
        "version": 2
      }
    }
[I 19:50:20.992 LabApp] JupyterLab extension loaded from /opt/conda/lib/python3.7/site-packages/jupyterlab
[I 19:50:20.992 LabApp] JupyterLab application directory is /opt/conda/share/jupyter/lab
[D 19:50:21.092 LabApp] Using default logger
[D 19:50:21.093 LabApp] Using default logger
[D 19:50:21.093 LabApp] Using default logger
[D 19:50:21.093 LabApp] Using default logger
[D 19:50:21.093 LabApp] Using default logger
[I 19:50:21.097 LabApp] Serving notebooks from local directory: /home/jovyan/.jupyter
[I 19:50:21.097 LabApp] The Jupyter Notebook is running at:
[I 19:50:21.097 LabApp] http://b832f09e7bfa:8889/?token=34c3ce50874939ad1fc2cc19e15f579329ddaf66dd9e0f90
[I 19:50:21.097 LabApp]  or http://127.0.0.1:8889/?token=34c3ce50874939ad1fc2cc19e15f579329ddaf66dd9e0f90
[I 19:50:21.097 LabApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
[C 19:50:21.103 LabApp] </pre>
</details>

<details><summary>Browser Output</summary>
<pre>
console.js:34 LSP:  Will invoke completer after .
completion.js:121 [LSP][Completer] Token: {offset: 2, value: ".", type: undefined}
completion.js:189 [LSP][Completer] Merging completions: {start: 3, end: 3, matches: Array(1126), metadata: {…}} {start: 0, end: 3, matches: Array(500), metadata: {…}}
completion.js:203 [LSP][Completer] Removing kernel prefix:  np.
</pre>

No errors

</details>

Thanks for help!

**Comment by ajacker** [comment]
me too, it usually takes long time

**Comment by krassowski** [comment]
Does it take longer for you than in another editor using the same language server (e.g. vim with pyls/VS Code with pyls), or is this the same in all editors?

**Comment by jhettler** [comment]
Hi @krassowski, it takes longer even in VS code with pyls, but just for the first time, if you try to write it again it is immediately autocompleted, but not in Jupyterlab. It is a bit faster in Jupyterlab after first try, but not as in VS Code.

**Comment by ajacker** [comment]
Hi, @krassowski ,i deploy it at my remote server (ubuntu 18.08) at the first time, it's too slow, about 10s+ for a hint. I thought it might be the network problem. But I deploy it at my PC(win10), it's still very slow(few secs), but better then my server.

**Comment by krassowski** [comment]
Thank you for your feedback. I guess it might be hardware dependent but then I did see very slow completion of one laptop which had good specs but was running Windows. Maybe we could make use of caching in a way?

Please feel free to suggest your own solution ideas!

**Comment by Mbarki66** [comment]
@krassowski First of all, great extension! Not perfect yet, but never thought all these features would be possible in jupyter notebooks!

The slow autocomplete is the only problem I encountered (on windows). I tried turning the jupyter-lsp extension off and as a result the autocomplete was much faster (using tab key of course). When I turn on the extension, the autocomplete becomes very slow again. Hope there is a solution to this.

**Comment by AMMJ93** [comment]
Have the same issue sadly, so I understand it is not user related?

**Comment by sntgluca** [comment]
Hi,

I wish to share my experience here.
I do experience similar issues, and the following configuration is the only one I could come up with which gives a sufficient performance:

```
jupyter-lsp 1.0.0
jupyterlab  2.0.2  
jedi             0.15.2 
parso         0.5.2
```

Any upgrade to newer versions of these packages, i.e. `jupyterlab==2.1.2` lead to backend errors slowing everything down, such as:

```
  File "<CONDA_ENV_DIR>/lib/python3.6/site-packages/parso/utils.py
", line 109, in python_bytes_to_unicode
    return unicode(source, encoding, errors)
LookupError: unknown encoding: 0
```
I hope this helps further investigation!

@krassowski, let me add that this extension is really useful and I am grateful it's available. Honestly, with JupyterLab it's still a bit laggy, I am sure it could get better, I hope you can continue improving it!

**Comment by aisensiy** [comment]
I have the same issue, seems slower then the default autocomplete...

**Comment by bergkvist** [comment]
I have the same problem. Running JupyterLab on Linux here (i7 6700K), and autocomplete is taking 5-10 seconds.

**Comment by krassowski** [comment]
There are two things that we can do to help:
- [ ] implement completionItem/resolve so that the documentation does not have to be computed and sent over for every possible completion
- [ ] on Linux use the ramdisk (when available) for the virtual documents

**Comment by krassowski** [comment]
In addition to the list above (now updated to four ideas for tackling this issue), there is an upstream issue for recent pyls versions: https://github.com/palantir/python-language-server/issues/823

It might be that some speed up will be seen once the upstream issue is resolved. If you are tracking this issue and would like to help but don't feel like learning typescript, head over to the [`python-language-server`](https://github.com/palantir/python-language-server) repo - it is pure Python :)

**Comment by quantum-booty** [comment]
My completion is also pretty slow, on top of that dot does not trigger automatic completion. Is there a setting to switch this on? Also pressing tab does not work for dot, one char, but works for two or more characters:
np. <tab> = no completion
np.a <tab> = no completion
np.ar <tab> = yes completion

**Comment by krassowski** [comment]
@quantum-booty the continuous completion (a.k.a hinterland) is an opt-in feature and you need to enable it in the Advanced Settings Editor → Code Completion:

![continues_hinting_and_docs](https://user-images.githubusercontent.com/5832902/92182596-fb074a80-ee43-11ea-8907-a842a9756d39.gif)

I suspect that you do not see auto-completion (i.e. completion just after typing dot) for `np.` just because there are so many candidate completions (599 when I just tried! numpy exports a huge namespace - though the 599 also includes kernel completions in it) that the requested bails before it is fully loaded.

Now to explain things a bit, it is not a surprise that the completions load slower with LSP:
- we load both the kernel and LSP server completions so we are limited by whichever is slower for any particular request
- the LSP completions carry the documentation which is heavy to transfer
   - some more advanced serves do not load documentation upfront but only after any of the completion items are activated; I plan to enable this for the servers which support it (e.g. R language server), but it has to wait until JupyterLab 3.0 with changes that were required upstream to expose the completion renderer to extensions is released.
- the first time the pyls (the Python server we recommend) loads completions for any specific package it has to analyse it which adds a bit overhead; the subsequent requests are faster as the underlying analyse (jedi) caches the results; however, as mentioned above there seems to be an upstream issue with the cache in the more recent versions of pyls.

Sorry that the experience is far from perfect! I do my best (with the great help of Nick!) but at the end of the day we are just doing this in our free time, and cannot move faster (although we would very much welcome any help!)

**Comment by sntgluca** [comment]
Hi @krassowski ,
just wish to restate that your work is great and the outcome is much appreciated.

From my side, I just enjoy it and try to give feedback to make sure things get even better...

**Comment by krassowski** [comment]
Thanks, @sntgluca! The feedback with a reproducible example like the one from @quantum-booty helps to narrow down what is the root cause fo the issues.

As for your comment on difficulty upgrading and parso errors - this is not something that has to do with this extension but with the pyls language server. they just had a new release hopefully fixing various errors (though there are some known issues with completion caching and diagnostics) - have a try upgrading and if you encounter any problems please do open a separate issue. I cannot help without detailed environment information.

**Comment by krassowski** [comment]
I did some profiling today on the `np.<tab>` case, which brings up about 600 suggestions.

With the material icons enabled the action took ~3.5s, while with the icons turned off it took ~2.5s.

### Icons
![Screenshot from 2020-09-18 19-03-18](https://user-images.githubusercontent.com/5832902/93630545-18125080-f9e2-11ea-8d85-d8b38fd5364b.png)

### No icons
![Screenshot from 2020-09-18 18-58-40](https://user-images.githubusercontent.com/5832902/93630571-1f395e80-f9e2-11ea-9dd4-1b8cd473646b.png)

| icons | no icons |
|--------|-------|
| ![Screenshot from 2020-09-18 19-04-40](https://user-images.githubusercontent.com/5832902/93630695-4a23b280-f9e2-11ea-9933-e2b187440cd6.png)  |  ![Screenshot from 2020-09-18 18-59-14](https://user-images.githubusercontent.com/5832902/93630667-3f691d80-f9e2-11ea-9c69-a2ce546e5712.png) |

Even with icons disabled 0.77s is used for the actual rendering. The rendering code is upstream.

There are different approaches we could use:
 - only display some of the completions
 - use windowing
 - improve how icons are generated upstream (to avoid generating hundreds of SVG objects)

**Comment by krassowski** [comment]
Would anyone here like to try running the [jedi cache script](https://github.com/krassowski/jupyterlab-lsp/blob/master/scripts/jedi_cache.py) and see if it makes a difference? Please adjust `MODULES_TO_CACHE` to include the modules you work with, and please let us know if this improved your experience, and which version of python-language-server and jedi you have installed.

**Comment by quantum-booty** [comment]
```
default jedi environment <SameEnvironment: 3.8.5 in /home/henryw/virtualenvs/data_sci>
jedi environment <jedi.api.environment.InterpreterEnvironment object at 0x7f1236173130>
jedi cache already exists, aborting warm up!
```

**Comment by quantum-booty** [comment]
![0](https://user-images.githubusercontent.com/28958399/93934699-7493a980-fd1b-11ea-8b9f-8510eddf55ad.jpg)
I deleted the jedi cache manually and ran the script, however it did not improve the speed. For numpy, the autocompete is still playing catch up after I finished typing np.array(). As you can see on the image, the autocompletion shows up after the documentation. It takes about 3 to 4 seconds for the autocompletion to appear.

**Comment by sntgluca** [comment]
Hi @krassowski ,
I wish to continue on this thread with something borderline with the topic.

In my case I came to the conclusion that lsp autocompletion is strongly bound to I/O and networking (remote NFS + jupyterhub proxy + VPN) and I don't think there's much to do to drastically improve that...
On the other hand the "native" autocompletion is much faster, including if I attach a kernel to the editor.

I am thinking of some potential workarounds which could work for cases like mine:
* Is it possible to _disable autocompletion_ and let the "native" autocompletion run instead?
* Have you been thinking of alternatives to jedi? Maybe I am too naive but could you think of starting a kernel automatically and bind the editor to that kernel?

Looking forward to your opinion

**Comment by sntgluca** [comment]
FYI,
I attempted rebuilding the cache with your script, but I only saw minor improvements (still taking 4-5 seconds to autocomplete).

Additionally I noticed a major difference with your snapshots above: in my case 'parsing' the instructions - as shown with highlighting - is really laggy and can take more than one second per statement. No error indication in the JS console. Just plain slow.

Hope it helps as a data point :)

**Comment by quantum-booty** [comment]
Could disabling completion on dot improve performance? Maybe add a setting to trigger completion 1 or 2 or 3 characters after the dot.

**Comment by bergkvist** [comment]
Or how about loading completion suggestions in multiple batches?

Quickly displaying the first ones without having to wait for everything. (And then more suggestions could be loaded in the background as you scroll down)

**Comment by quantum-booty** [comment]
Does pyls calculate or record the function call statistics? Maybe keep a shortlist of the recently used function to make them more readily available for the next autocompletion.

**Comment by quantum-booty** [comment]
@krassowski This might be a stupid question, but why not use jedi for completion, and pyls for the other lsp features? Pyls uses jedi under the hood anyways, but apparently with a lot of inefficiencies when it comes to completion.

**Comment by krassowski** [comment]
@bergkvist yes, we could try this using `TriggerForIncompleteCompletions` and `PartialResultParams` from https://microsoft.github.io/language-server-protocol/specification. On the frontend I think that we need to give up on re-using the built-in completer and create a custom one for this extension which will emit appropriate signals...

**Comment by krassowski** [comment]
@quantum-booty you might be interested in checking out https://github.com/pappasam/jedi-language-server and using it instead of pyls. This is basically a language server for Python using mostly jedi stuff and might be quicker than pyls. You would need to create a language spec familiar to those in [jupyter_lsp/specs](https://github.com/krassowski/jupyterlab-lsp/blob/master/py_src/jupyter_lsp/specs/) and have a look at [CONTRIBUTING.md#specs](https://github.com/krassowski/jupyterlab-lsp/blob/master/CONTRIBUTING.md#specs). Some bits might be outdated so shut as a message if you need further guidaince.

**Comment by quantum-booty** [comment]
@krassowski Is something like this fine? What do I do from here?
![image](https://user-images.githubusercontent.com/28958399/94350156-6f986800-0043-11eb-9934-dcffa006148d.png)

**Comment by badiku** [comment]
really slow. it takes more than 10 seconds to show the autocompletion, after tying dot, even if it's a little module with 20 lines

**Comment by sntgluca** [comment]
Hi @krassowski  and @quantum-booty ,

I am able to work with `jedi-language-server` and the performance is indeed much better. I feel satisfied with the result. Loading at the beginning is a bit slow (1-2 seconds to fill up the content for pandas for instance) but every other lookup is pretty fast.

My manual setup is as follows:
* install jedi-language-server via pip (in a conda environment in my case)
* edit  jupyter_notebook.py as discussed in https://jupyterlab-lsp.readthedocs.io/en/latest/Configuring.html#installing-language-servers
```python
c.LanguageServerManager.language_servers = {
    "jedi-language-server": {
        "argv": ["jedi-language-server"],
        "languages": ["python"],
        "version": 2
    }
}
c.LanguageServerManager.autodetect = False
```

I do have other problems, including the fact that diagnostics is not working, but this is an enabler and I am really happy to have a (partial) solution!

**Comment by krassowski** [comment]
You may benefit from the new [2.1.2 release](https://github.com/krassowski/jupyterlab-lsp/releases/tag/v2.1.2): in my profiling I found that one of the bottlenecks for fast completion, especially in continuous hinting mode, was actually not the completion itself but the highlights feature implementation. Prior to this release the request for highlights was sent on every typed character and every cursor movement which was causing a "traffic jam" for the pyls language server (it was not fast enough with providing highlights so the subsequent requests - such as completions - needed to wait); this was especially a problem for long files/notebooks with many variables. In this new release improvements were made:
- only sending the request for highlights if the token under the cursor changed
-  not sending straight away but debouncing, which reduces the number of request sent when typing (default `debouncerDelay` is set to 250 ms, please adjust for your typing speed)
- a few other bug fixes to the highlights and completion features

Please let me know if you see an improvement.

**Comment by shooney0** [comment]
Firstly Thank you for an extremely helpful extension.

I do seem to experience slow completion on windows. 
I can retrieve an array of 134 types in 3 seconds which is still impressive to me.

This is by no means something i'm familiar with however I was wondering if this is working as intended or if I have caused this in some way.

![image](https://user-images.githubusercontent.com/9410843/103473012-0d8c9300-4de8-11eb-9876-c7e94646f7e1.png)

During the typing out with auto completion, the Chrome browser console waits for all Fetching and Transforming requests to resolve all their merging completions in queue before returning results. Is this a bottle neck that is negatively impacting the speed of the results returning?

Currently running
jedi                   0.17.2
jupyterlab         2.2.9
jupyter-lsp            0.9.3
python-language-server 0.36.2
@jupyter-widgets/jupyterlab-manager v2.0.0 enabled  ok
@krassowski/jupyterlab-lsp v2.1.2 enabled  ok

**Comment by krassowski** [comment]
> During the typing out with auto completion, the Chrome browser console waits for all Fetching and Transforming requests to resolve all their merging completions in queue before returning results. Is this a bottle neck that is negatively impacting the speed of the results returning?

I guess that we could abort early if a new request was sent but it is not the bottlneck; if you can optimize the code, you are welcome to send a PR.

**Comment by sntgluca** [comment]
Dear @krassowski ,

unfortunately the latest changes don't seem to have improved much. I have to admit that compared to the "native" autocomplete experience in the notebooks, driven by a running kernel, lsp is still lagging behind.

It's hard to understand, with my lack of technical expertise, why there's such a big difference. But I ask if the bottleneck could outside your plugin, either in the I/O with the pyls / jedi servers or in these servers themselves.
When I use autocomplete, I see my system's CPU usage spike to full power for several seconds, which could be an indication of .

I hope you don't mind if I reiterate some points @quantum-booty and I shared above independently, which I think are still valid:
* the lsp plugin has several interesting features, not only autocomplete. Would it be possible to _disable_ autocomplete selectively so we can still benefit from the other features (PEP8 errors, highlights and so on) while using the native autocomplete?
* could you consider hooking the autocomplete logic of lsp to a running ipykernel process, at least to assess lsp's performance? Would it be a lot of work?

I don't have the technical expertise to help, unfortunately, yet I think this plugin is our best chance to closing the gap between JupyterLab and other IDEs. I hope you can explore these ideas and I would like continuing testing your progress.

**Comment by krassowski** [comment]
@sntgluca yes, what you describe is partially tracked in #440.

Yes, the bottlneck is certainly in part in the pyls because from the reports above using another language server (jedi-language-server) results in faster autocompletion. There is for example this advice on pyls autocompletion speed on pyls issue tracker: https://github.com/palantir/python-language-server/issues/867#issuecomment-707116310.

**Comment by krassowski** [comment]
Support for `completionItem/resolve` is [basically ready](https://github.com/krassowski/jupyterlab-lsp/pull/487), but neither `pyls` nor `jedi-language-server` implements it; if you would like to help it might be a good opportunity (only Python skills required). Same for Julia language server (only Julia required). It is not clear which language server projects will be happy about implementing it, but hopefully most of them will accept PRs (please do check with them if you intend on contribution and do keep us in the loop!).

**Comment by krassowski** [comment]
Good news!

1. Performance-focused improvements were made to the asynchronous completions logic in v3.3.0:
   - we do not wait for the kernel response forever; instead `kernelResponseTimeout` (by default 0.6s) is used; if kernel is too slow or busy, we will show the LSP completions only; you can adjust the timeout or disable it completely.
   - you can opt-in to not wait for the kernel if it is busy (with computations) by toggling `waitForBusyKernel` setting
   - delayed completions are now rejected if the cursor has moved (e.g. to another line) improving the UX and reducing the number of layout reflow operations on the frontend
   - the server extension (v1.1.2) now works reliably with long LSP messages (including completions for large packages), especially on Windows; the maximum Windows sleep on idle stream was reduced from 2 s to 0.1 s on Windows (and increased from 0.05 s to 0.1 s on other systems)

2. I sent pull requests to `pyls` (https://github.com/palantir/python-language-server/pull/905) and `jedi-language-server` (https://github.com/pappasam/jedi-language-server/pull/56) implementing `completionItem/resolve`. I hope it won't take too long to review. In the meantime you can try them out for yourself.

    First get the latest versions:

    ```
    pip install -U jupyterlab==3 jupyterlab-lsp==3.3 jupyter-lsp==1.1.2
    ```

    Then install patched `pyls`:
    
    ```
    pip install git+https://github.com/krassowski/python-language-server.git@feature/completionItem/resolve
    ```

    or `jedi-language-server` (please note that it requires the new Jedi version that is not compatible with the latest IPython yet):
   
    ```
    pip install git+https://github.com/krassowski/jedi-language-server.git@feature/completionItem/resolve
    ```

It could really help if you could test these changes and provide us with before/after comparison (ideally in a new bug report issue).

**Comment by pappasam** [comment]
@krassowski your PR has been merged / released in jls `0.23.0`.

**Comment by krassowski** [comment]
After my pyls [profiling](https://github.com/palantir/python-language-server/issues/823#issuecomment-774695123) revealed that the remaining slowness for numpy and others is due to label formatting (precisely `Completion.get_signatures()`) I was working to workaround it (and improve documentation too).

There is a trade-off between getting the nice completion label with parameters and performance. I tried fetching it asynchronously but it leads to too many issues (jedi does not like to be run from multiple threads at once). In the current version:

- if there are many completions (>25 by default, configurable by `pyls.plugins.jedi_completion.resolve_at_most_label`) only the names instead of names + parameters (e.g. `add` instead of `add(a, b)`) are shown
- added a cache with time-based invalidation (30 minutes) which is only being used for the numpy, pandas, tenserflow, and matplotlib (following advice from the jedi master)
- moved a second call to `Completion.get_signatures()` behind a snippet conditional which helps as we do not use snippets.

If you wish to check it out you can use:

```bash
pip install -U jupyterlab==3 jupyterlab-lsp==3.3.1 jupyter-lsp==1.1.3
pip install git+https://github.com/krassowski/python-language-server.git@main
```

The other work in this fork concerns rendering documentation in the panel (and elsewhere) as proper markdown:

![Screenshot from 2021-02-08 00-50-33](https://user-images.githubusercontent.com/5832902/107164951-0fc6ac00-69a9-11eb-9408-1b377b569399.png)

If you have any issues with the fork, or if you still experience slowness, please upload the profiling `.prof` files which can be generated by running:

```bash
pip install pytest numpy
pytest -k test_numpy_completions
```

While this is still WIP, any testing and feedback will be greatly appreciated!

**Comment by krassowski** [comment]
In addition to the fork mentioned in https://github.com/krassowski/jupyterlab-lsp/issues/272#issuecomment-774806496, please use:

```
%config Completer.use_jedi = False
```

for IPython kernel to disable jedi completions from kernel. This will still give you keys of dictionaries/columns and paths and other kernel completions, but avoid repeating the jedi completions calculation which is expensive and is done better by the LSP.

When I combined the fork and the above I am getting a very pleasant experience. This is probably the way to go forward.

**Comment by fsondej** [comment]
Awesome, it works MUCH better after this https://github.com/krassowski/jupyterlab-lsp/issues/272#issuecomment-774806496 

One problem I noticed, is that completions block typing. You can reproduce it by typing `np.`, then wait for the completion to show, and then type `ra` fast. It looks like even after you have typed it all, it will first look for completions for `np.r`, then `np.ra`.
It's even more visible, when you type `np.`, wait for completion, type `random` and then hold backspace - it won't let you delete the whole thing, it will stop at each char to get the completion.

I tested this before the change, and it seems to happen too, but it was less noticeable, because the blocking only seems to happen if the completion box is already shown - and with completions working slower, usually it won't be shown quickly enough to cause a problem.

It would be better if typing a new character (or deleting) interrupted the completion.

**Comment by sntgluca** [comment]
Wow, @krassowski the improvement is impressive, I couldn't think it was even possible a few months 
 ago :)
My initial tests show this is now working well (in the German sense of the word).

That's great one can start thinking about further improvements such as the one suggested by @fsondej .
Thanks for the care you put into this.

**Comment by sntgluca** [comment]
BTW can you confirm this would only work on JupyterLab3?
I'd be curious to know about the amount of work needed to backport to JupyterLab2, as at the moment not every package is ready for the transition and sticking to Jupyterlab2 is still a necessity...

**Comment by krassowski** [comment]
> BTW can you confirm this would only work on JupyterLab3?

The major sources of performance improvements since the transition to JupyterLab 3.0 are:
1. the calls to resolve signatures are avoided for big queries and cached for numpy/pandas (in my fork)
2. the documentation retrieval is deferred until the item is selected in the completer (in my fork, PR sent upstream)
3. not waiting too long for kernel responses when it is slow (in 3.x version of extension)
4. in addition you can make the kernel even more responsive by disabling jedi in kernel manually as recommended in https://github.com/krassowski/jupyterlab-lsp/issues/272#issuecomment-775726255

Which means that:
- (1) and (4) will work and help in JupyterLab 2.x;
- (2) requires JupyterLab 3.0 and if used with JupyterLab 2.x you will not get this benefit (and you will need to set `pyls.plugins.jedi_completion.eager` to true to get documentation at all); this cannot be backported
- (3) is an extension change that I am reluctant to backport because it introduced a new setting so it will probably remain as JupyterLab 3.0 only feature

**Comment by bollwyvl** [comment]
@krassowski Good summary. 

@sntgluca Of course, PRs welcome, but at this point, such a notional contributor, capable of wading into _this_ codebase (here be :dragon:) may well be better off spending their time trying to get said _other extensions_ working in Lab 3. 

A lot of the PRs around the ecosystem to support 3 are quite straightforward, and improve everybody's lives (users and maintainers alike). If said extension _doesn't_ have a "Please Support JupyterLab 3" issue, it's great to get one started... if it does exist, a nice emoji on the header is good, but `mee tooo!!!111! NEED now! when will it be ready?!?!` comments don't help move things forward... we're all a little exhausted from this whole exercise.

**Comment by sntgluca** [comment]
Hi @bollwyvl , @krassowski ,
I hope I didn't sound pushy, it wasn't my intention.
Based on your answers I agree it's not worth it and efforts should be rather spent on moving forward.

**Comment by g6ai** [comment]
@krassowski Thanks for the brilliant work! I tried https://github.com/krassowski/jupyterlab-lsp/issues/272#issuecomment-775726255 and it works well for local kernels, faster autocompletion, impressive.

However, with remote ipython kernels I installed with e.g. [remote_ikernel](https://pypi.org/project/remote_ikernel/), there's no autocompletion.

**Comment by krassowski** [comment]
@g6ai please open a new [bug report](https://github.com/krassowski/jupyterlab-lsp/issues/new?assignees=&labels=&template=bug_report.md) and provide all the details including the debug output from kernel.

**Comment by bollwyvl** [comment]
Yeah, once you go remote, the topology will get weird between what we know about and what your language server knows about. We're not really ready for that, partially see:

https://github.com/krassowski/jupyterlab-lsp/issues/184#issuecomment-777682465

**Comment by chenshen03** [comment]
How do I turn off completion but keep the rest of jupyterlab-lsp working?

**Comment by autoscraper** [comment]
I also would like to know how do I turn off completion but keep the rest of jupyterlab-lsp working.
The completion experience have been awful. Lost many hours trying make it function properly, but always stop working at some point.

**Comment by sntgluca** [comment]
Hi @krassowski ,
I am experiencing some additional pyls configuration issues, which I consider in scope within this thread. I hope you could give me a hint.

It turns out pyls is constantly consuming more than 50% of CPU time, trying to index all python files in my home, which are plenty as I have several python files and separate environments, similarly to what is described in https://github.com/palantir/python-language-server/issues/421

Do you have any configuration suggestion to limit this behaviour?

**Comment by krassowski** [comment]
@sntgluca does it occur using my pyls fork, or the upstream? How do you know that it is indexing files? Is it coming from a specific pyls plugin? If this is an issue with pyls, then please open a detailed issue upstream, but chance is that it may be with a specific pyls plugin - then please open it in appropriate repository.

**Comment by sntgluca** [comment]
Dear @krassowski ,

I wanted to check with you first, maybe there was an obvious solution. Thanks for following up with me quickly anyway.
I opened a separate ticket: https://github.com/palantir/python-language-server/issues/928

And to answer your questions:

I have been using `strace -p $(pgrep pyls)` to monitor the files accessed by `pyls` as described in https://github.com/palantir/python-language-server/issues/421#issuecomment-514524655

I can confirm the same behaviour occurs both with your fork `python-language-server-0.36.2+49.g3536061` and with the official pip package `python-language-server==0.36.2`

I don't know which pyls plugin is involved or how to check it... sorry.

**Comment by krassowski** [comment]
Thanks! FYI, we will be moving towards the https://github.com/python-lsp/python-lsp-server fork soon (once I get to send PRs that provide the performance improvements). At very least, you can check which plugins are installed using `pip list | grep pyls`.

**Comment by droctothorpe** [comment]
https://github.com/krassowski/jupyterlab-lsp/issues/272#issuecomment-774806496 brought my completion time for a Dask Gateway `Gateway` object down from 5 seconds to less than 0.5 seconds! Thank you so much!

**Comment by bollwyvl** [comment]
So now that all of the primary upstream pyls dependencies support `python-lsp-server`, perhaps we can close this?

**Comment by sntgluca** [comment]
Hi!
Personally I think it's safe to close the ticket... the progress has been tremendous and the quality of the package is pretty high. Of course that's only my 2 cents.

Just one more note, please. Can you guys comment on the status of the https://github.com/python-lsp/python-lsp-server integration, and your current suggestions on the current best practices / configuration?

Thanks for all the help!

**Comment by krassowski** [comment]
No, none of my performance improvements got into pylsp yet. If anyone feels like picking up please do, I am happy to advise but at capacity right now.

**Comment by krassowski** [comment]
Though as a side note, one can now easily checkout the pyright server and compare performance as this one does not use jedi. From what I saw we are now hitting the point with pyright where it is our rendering and not the server that is the culprit of the delay (especially icon SVG parsing which is - ridiculously - done each time rather than only once per icon type).

**Comment by krassowski** [comment]
The big news is that it is now possible to disable completions from specific source (e.g. kernel or LSP servers) in the just released [3.7 version](https://github.com/krassowski/jupyterlab-lsp/releases/tag/v3.7.0).

**Comment by potoo0** [comment]
> The big news is that it is now possible to disable completions from specific source (e.g. kernel or LSP servers) in the just released [3.7 version](https://github.com/krassowski/jupyterlab-lsp/releases/tag/v3.7.0).

How to disable lsp for some specific language, and use kernel completions.
I want to use metals for scala, and disable pylsp for python.

**Comment by sntgluca** [comment]
Hi,

I wish to share one additional note which I think could be of interest:
I really got optimal results with @krassowski 's patched version of python-language-server, while using the default recommended python-lsp-server autocomplete is still pretty lagging.

This means, for me, that you're doing a wonderful job, but without integrating the latest features to the downstream packages, we won't get all the benefits. I'd love to be able to help out but I am not good enough to contribute...

**Comment by krassowski** [comment]
> How to disable lsp for some specific language, and use kernel completions.
> I want to use metals for scala, and disable pylsp for python.

Hi @potoo0 sorry I missed your message earlier. You can disable completion for pylsp **only** by modifying its settings:

```json
{
    "language_servers": {
        "pylsp": {
            "serverSettings": {
                "pylsp.plugins.jedi_completion.enabled": false,
                "pylsp.plugins.rope_completion.enabled": false
            }
        }
    }
}
```

@sntgluca it's on my todo-list later this month.

**Comment by krassowski** [comment]
Quick update: the two pull requests with major performance improvements to [python-lsp-server](https://github.com/python-lsp/python-lsp-server) were merged yesterday. With the default settings you should get a substantial speed up. You can try it out by installing directly from GitHub:

```
pip install -U jupyterlab jupyterlab-lsp nbclassic
pip uninstall python-language-server python-lsp-server
pip install git+https://github.com/python-lsp/python-lsp-server.git@develop
```

There were also some improvements for the rendering of icons in #625.

**Comment by krassowski** [comment]
`python-lsp-server` 1.2 is now available. I will close this issue for now as I believe it was solved - numerous performance improvements were added since it was created and most crucially the real culprit, this is very slow calls to `jedi` in the default pyls server has been eliminated.

For anyone reading this comment in future, please note that the fixes are available only in my personal `pyls` fork and in the community `pylsp` fork, but not in the original `pyls` which is no longer maintained (so you need `python-lsp-server`, not `python-language-server`); in short:

```bash
pip uninstall python-language-server
pip install -U jupyterlab jupyterlab-lsp nbclassic python-lsp-server
```

**Comment by hellojinwoo** [comment]
@krassowski Would it be better if I `uninstall python-language-server` and `install -U jupyterlab jupyterlab-lsp nbclassic python-lsp-server`, when I use your `jupyter-lsp` on `jupyter desktop` as well? In other words, is this needed for `Jupyterlab Desktop` also?
